xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -1042,6 +1193,114 @@ private void sendNackRequest(String typeUrl, Collection<String> resourceNames,
     }
   }
 
+  /**
+   * A task handle that can schedule, cancel and reschedule the task for multiple times.
+   */
+  private final class ReschedulableTaskHandle {
+    private final Runnable task;
+    @Nullable
+    private ScheduledHandle scheduledHandle;
+
+    ReschedulableTaskHandle(Runnable task) {
+      this.task = task;
+    }
+
+    void schedule(long delay, TimeUnit unit) {
+      checkState(scheduledHandle == null || !scheduledHandle.isPending(),
+          "Task already scheduled and not finished.");
+      scheduledHandle = syncContext.schedule(task, delay, unit, timeService);
+    }
+
+    void cancel() {
+      if (scheduledHandle != null) {
+        scheduledHandle.cancel();
+      }
+    }
+  }
+
+  private abstract class ResourceFetchTimeoutTask implements Runnable {
+    protected final String resourceName;
+
+    ResourceFetchTimeoutTask(String resourceName) {
+      this.resourceName = resourceName;
+    }
+
+    ReschedulableTaskHandle toTaskHandle() {
+      return new ReschedulableTaskHandle(this);
```

Comment:
You mean something like `new ReschedulableTaskHandle(new CdsResourceFetchTimeoutTask(...))`? I was trying to avoid this kind of nested construction, it's inelegant in terms of readability.

Reply:
I think only nest once is not a big deal.

```
new ReschedulableTaskHandle(
    new CdsResourceFetchTimeoutTask(cdsResourceName))
```

or
```java
Runnable cdsResourceFetchTimeoutTask = new CdsResourceFetchTimeoutTask(cdsResourceName);
new ReschedulableTaskHandle(cdsResourceFetchTimeoutTask);
```

both look normal.

If you move the `CdsResourceFetchTimeoutTask` class definition to a local class inside the function, you don't even need to introduce the abstract class `ResourceFetchTimeoutTask` and the string arg constructor. (However, in test you need use constant string "CdsResourceFetchTimeoutTask" for the filter)

```
final String cdsResourceName;
class CdsResourceFetchTimeoutTask implements Runnable {
  @Override
  public void run() {...}
};
new ReschedulableTaskHandle(new CdsResourceFetchTimeoutTask());
```

Reply:
The main purpose of `ResourceFetchTimeoutTask` is to provide the common implementation of `toHandler()`, which gives a smooth transition from a task to a task handle. 

Also, it groups `{Lds/Rds/Cds/Eds}ResourceFetchTimeoutTask` classes together such that the `resourceName` field only needs to be defined in the super class (instead of each class defining its own `resourceName`).

Reply:
As discussed offline, we will simply reschedule resource fetch timeout task for all resources after RPC retry, no matter if the result is known before the retry or not.